[0.12] Panic on shutdown

Got a random panic when I tried to shutdown my local instance:

```
monitor] 2016/03/28 16:56:33 shutting down monitor system
[monitor] 2016/03/28 16:56:33 terminating storage of statistics
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x30 pc=0x2325c5]

goroutine 300686 [running]:
github.com/influxdata/influxdb/tsdb.(*Shard).CreateIterator(0xc841b8cfc0, 0x1907a08, 0xc82dd29aa0, 0xbed000, 0x0, 0x0, 0xc84a6aa460, 0x1, 0x1, 0x0, ...)
	/Users/michaeldesa/go/src/github.com/influxdata/influxdb/tsdb/shard.go:420 +0x115
github.com/influxdata/influxdb/tsdb.(*shardIteratorCreator).CreateIterator(0xc86fe646a0, 0x1907a08, 0xc82dd29aa0, 0xbed000, 0x0, 0x0, 0xc84a6aa460, 0x1, 0x1, 0x0, ...)
	/Users/michaeldesa/go/src/github.com/influxdata/influxdb/tsdb/shard.go:886 +0x69
github.com/influxdata/influxdb/influxql.IteratorCreators.CreateIterator.func1(0xc8200c8a00, 0x12, 0x20, 0xc8b1630990, 0xc8b1630880, 0x0, 0x0)
	/Users/michaeldesa/go/src/github.com/influxdata/influxdb/influxql/iterator.go:483 +0xdf
github.com/influxdata/influxdb/influxql.IteratorCreators.CreateIterator(0xc8200c8a00, 0x12, 0x20, 0x1907a08, 0xc82dd29aa0, 0xbed000, 0x0, 0x0, 0xc84a6aa460, 0x1, ...)
	/Users/michaeldesa/go/src/github.com/influxdata/influxdb/influxql/iterator.go:481 +0xd8
github.com/influxdata/influxdb/influxql.(*IteratorCreators).CreateIterator(0xc8331b5f40, 0x1907a08, 0xc82dd29aa0, 0xbed000, 0x0, 0x0, 0xc84a6aa460, 0x1, 0x1, 0x0, ...)
	<autogenerated>:152 +0xeb
github.com/influxdata/influxdb/influxql.buildExprIterator.func1(0xc82dd29aa0, 0x1907e28, 0xc8331b5f40, 0xc8b1631240, 0x0, 0x0, 0x0, 0x0)
	/Users/michaeldesa/go/src/github.com/influxdata/influxdb/influxql/select.go:255 +0x196
github.com/influxdata/influxdb/influxql.buildExprIterator(0x1907a08, 0xc82dd29aa0, 0x1907e28, 0xc8331b5f40, 0x1907a08, 0xc82dd29aa0, 0xbed000, 0x0, 0x0, 0xc84a6aa460, ...)
	/Users/michaeldesa/go/src/github.com/influxdata/influxdb/influxql/select.go:339 +0x13a0
github.com/influxdata/influxdb/influxql.buildFieldIterators.func1(0xc86fe64618, 0x1, 0x1, 0x1907e28, 0xc8331b5f40, 0xc8b16315d8, 0xc84a6aa840, 0x1, 0x1, 0x0, ...)
	/Users/michaeldesa/go/src/github.com/influxdata/influxdb/influxql/select.go:154 +0x5b6
github.com/influxdata/influxdb/influxql.buildFieldIterators(0xc86fe64618, 0x1, 0x1, 0x1907e28, 0xc8331b5f40, 0x0, 0x0, 0xbed000, 0x0, 0x0, ...)
	/Users/michaeldesa/go/src/github.com/influxdata/influxdb/influxql/select.go:140 +0xf8
github.com/influxdata/influxdb/influxql.Select(0xc8861ce000, 0x1907e28, 0xc8331b5f40, 0xc8b1631c90, 0x0, 0x0, 0x0, 0x0, 0x0)
	/Users/michaeldesa/go/src/github.com/influxdata/influxdb/influxql/select.go:81 +0xba9
github.com/influxdata/influxdb/cluster.(*QueryExecutor).executeSelectStatement(0xc8200a4400, 0xc8861ce000, 0x2710, 0x0, 0xc82ec42480, 0xc82ec424e0, 0x0, 0x0)
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -32,6 +32,7 @@
 - [#6092](https://github.com/influxdata/influxdb/issues/6092): Upgrading directly from 0.9.6.1 to 0.11.0 fails
 - [#6061](https://github.com/influxdata/influxdb/issues/6061): [0.12 / master] POST to /write does not write points if request has header 'Content-Type: application/x-www-form-urlencoded'
 - [#6121](https://github.com/influxdata/influxdb/issues/6121): Fix panic: slice index out of bounds in TSM index
+- [#6140](https://github.com/influxdata/influxdb/issues/6140): Ensure Shard engine not accessed when closed.
 
 ## v0.11.0 [2016-03-22]
 

--- a/tsdb/shard.go
+++ b/tsdb/shard.go
@@ -76,11 +76,11 @@ type Shard struct {
 	database        string
 	retentionPolicy string
 
-	engine  Engine
 	options EngineOptions
 
 	mu                sync.RWMutex
 	measurementFields map[string]*MeasurementFields // measurement name to their fields
+	engine            Engine
 
 	// expvar-based stats.
 	statMap *expvar.Map
@@ -217,6 +217,9 @@ type SeriesCreate struct {
 
 // WritePoints will write the raw data points and any new metadata to the index in the shard
 func (s *Shard) WritePoints(points []models.Point) error {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+
 	s.statMap.Add(statWriteReq, 1)
 
 	seriesToCreate, fieldsToCreate, seriesToAddShardTo, err := s.validateSeriesAndFields(points)
@@ -255,9 +258,7 @@ func (s *Shard) WritePoints(points []models.Point) error {
 			}
 
 			// This was populated earlier, don't need to validate that it's there.
-			s.mu.RLock()
 			mf := s.measurementFields[p.Name()]
-			s.mu.RUnlock()
 
 			// If a measurement is dropped while writes for it are in progress, this could be nil
 			if mf == nil {
@@ -284,6 +285,8 @@ func (s *Shard) WritePoints(points []models.Point) error {
 
 // DeleteSeries deletes a list of series.
 func (s *Shard) DeleteSeries(seriesKeys []string) error {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
 	return s.engine.DeleteSeries(seriesKeys)
 }
 
@@ -307,8 +310,8 @@ func (s *Shard) createFieldsAndMeasurements(fieldsToCreate []*FieldCreate) (map[
 		return nil, nil
 	}
 
-	s.mu.Lock()
-	defer s.mu.Unlock()
+	s.mu.RLock()
+	defer s.mu.RUnlock()
 
 	// add fields
 	measurementsToSave := make(map[string]*MeasurementFields)
@@ -387,10 +390,16 @@ func (s *Shard) validateSeriesAndFields(points []models.Point) ([]*SeriesCreate,
 }
 
 // SeriesCount returns the number of series buckets on the shard.
-func (s *Shard) SeriesCount() (int, error) { return s.engine.SeriesCount() }
+func (s *Shard) SeriesCount() (int, error) {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+	return s.engine.SeriesCount()
+}
 
 // WriteTo writes the shard's data to w.
 func (s *Shard) WriteTo(w io.Writer) (int64, error) {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
 	n, err := s.engine.WriteTo(w)
 	s.statMap.Add(statWriteBytes, int64(n))
 	return n, err
@@ -401,6 +410,8 @@ func (s *Shard) CreateIterator(opt influxql.IteratorOptions) (influxql.Iterator,
 	if influxql.Sources(opt.Sources).HasSystemSource() {
 		return s.createSystemIterator(opt)
 	}
+	s.mu.RLock()
+	defer s.mu.RUnlock()
 	return s.engine.CreateIterator(opt)
 }
 
@@ -472,6 +483,8 @@ func (s *Shard) SeriesKeys(opt influxql.IteratorOptions) (influxql.SeriesList, e
 		return []influxql.Series{{Aux: auxFields}}, nil
 	}
 
+	s.mu.RLock()
+	defer s.mu.RUnlock()
 	return s.engine.SeriesKeys(opt)
 }
 

--- a/tsdb/shard.go
+++ b/tsdb/shard.go
@@ -42,6 +42,10 @@ var (
 	// ErrFieldUnmappedID is returned when the system is presented, during decode, with a field ID
 	// there is no mapping for.
 	ErrFieldUnmappedID = errors.New("field ID not mapped")
+
+	// ErrEngineClosed is returned when a caller attempts indirectly to
+	// access the shard's underlying engine.
+	ErrEngineClosed = errors.New("engine is closed")
 )
 
 // A ShardError implements the error interface, and contains extra
@@ -181,6 +185,14 @@ func (s *Shard) close() error {
 	return err
 }
 
+// closed determines if the Shard is closed.
+func (s *Shard) closed() bool {
+	s.mu.RLock()
+	closed := s.engine == nil
+	s.mu.RUnlock()
+	return closed
+}
+
 // DiskSize returns the size on disk of this shard
 func (s *Shard) DiskSize() (int64, error) {
 	stats, err := os.Stat(s.path)
@@ -217,9 +229,9 @@ type SeriesCreate struct {
 
 // WritePoints will write the raw data points and any new metadata to the index in the shard
 func (s *Shard) WritePoints(points []models.Point) error {
-	s.mu.RLock()
-	defer s.mu.RUnlock()
-
+	if s.closed() {
+		return ErrEngineClosed
+	}
 	s.statMap.Add(statWriteReq, 1)
 
 	seriesToCreate, fieldsToCreate, seriesToAddShardTo, err := s.validateSeriesAndFields(points)
@@ -258,7 +270,9 @@ func (s *Shard) WritePoints(points []models.Point) error {
 			}
 
 			// This was populated earlier, don't need to validate that it's there.
+			s.mu.RLock()
 			mf := s.measurementFields[p.Name()]
+			s.mu.RUnlock()
 
 			// If a measurement is dropped while writes for it are in progress, this could be nil
 			if mf == nil {
@@ -285,22 +299,26 @@ func (s *Shard) WritePoints(points []models.Point) error {
 
 // DeleteSeries deletes a list of series.
 func (s *Shard) DeleteSeries(seriesKeys []string) error {
-	s.mu.RLock()
-	defer s.mu.RUnlock()
+	if s.closed() {
+		return ErrEngineClosed
+	}
 	return s.engine.DeleteSeries(seriesKeys)
 }
 
 // DeleteMeasurement deletes a measurement and all underlying series.
 func (s *Shard) DeleteMeasurement(name string, seriesKeys []string) error {
-	s.mu.Lock()
-	defer s.mu.Unlock()
+	if s.closed() {
+		return ErrEngineClosed
+	}
 
 	if err := s.engine.DeleteMeasurement(name, seriesKeys); err != nil {
 		return err
 	}
 
 	// Remove entry from shard index.
+	s.mu.Lock()
 	delete(s.measurementFields, name)
+	s.mu.Unlock()
 
 	return nil
 }
@@ -310,8 +328,8 @@ func (s *Shard) createFieldsAndMeasurements(fieldsToCreate []*FieldCreate) (map[
 		return nil, nil
 	}
 
-	s.mu.RLock()
-	defer s.mu.RUnlock()
+	s.mu.Lock()
+	defer s.mu.Unlock()
 
 	// add fields
 	measurementsToSave := make(map[string]*MeasurementFields)
@@ -347,9 +365,6 @@ func (s *Shard) validateSeriesAndFields(points []models.Point) ([]*SeriesCreate,
 	var seriesToAddShardTo []string
 
 	// get the shard mutex for locally defined fields
-	s.mu.RLock()
-	defer s.mu.RUnlock()
-
 	for _, p := range points {
 		// see if the series should be added to the index
 		if ss := s.index.Series(string(p.Key())); ss == nil {
@@ -363,7 +378,10 @@ func (s *Shard) validateSeriesAndFields(points []models.Point) ([]*SeriesCreate,
 		}
 
 		// see if the field definitions need to be saved to the shard
+		s.mu.RLock()
 		mf := s.measurementFields[p.Name()]
+		s.mu.RUnlock()
+
 		if mf == nil {
 			for name, value := range p.Fields() {
 				fieldsToCreate = append(fieldsToCreate, &FieldCreate{p.Name(), &Field{Name: name, Type: influxql.InspectDataType(value)}})
@@ -391,15 +409,17 @@ func (s *Shard) validateSeriesAndFields(points []models.Point) ([]*SeriesCreate,
 
 // SeriesCount returns the number of series buckets on the shard.
 func (s *Shard) SeriesCount() (int, error) {
-	s.mu.RLock()
-	defer s.mu.RUnlock()
+	if s.closed() {
+		return 0, ErrEngineClosed
+	}
 	return s.engine.SeriesCount()
 }
 
 // WriteTo writes the shard's data to w.
 func (s *Shard) WriteTo(w io.Writer) (int64, error) {
-	s.mu.RLock()
-	defer s.mu.RUnlock()
+	if s.closed() {
+		return 0, ErrEngineClosed
+	}
 	n, err := s.engine.WriteTo(w)
 	s.statMap.Add(statWriteBytes, int64(n))
 	return n, err
@@ -407,11 +427,13 @@ func (s *Shard) WriteTo(w io.Writer) (int64, error) {
 
 // CreateIterator returns an iterator for the data in the shard.
 func (s *Shard) CreateIterator(opt influxql.IteratorOptions) (influxql.Iterator, error) {
+	if s.closed() {
+		return nil, ErrEngineClosed
+	}
+
 	if influxql.Sources(opt.Sources).HasSystemSource() {
 		return s.createSystemIterator(opt)
 	}
-	s.mu.RLock()
-	defer s.mu.RUnlock()
 	return s.engine.CreateIterator(opt)
 }
 
@@ -468,6 +490,10 @@ func (s *Shard) FieldDimensions(sources influxql.Sources) (fields, dimensions ma
 
 // SeriesKeys returns a list of series in the shard.
 func (s *Shard) SeriesKeys(opt influxql.IteratorOptions) (influxql.SeriesList, error) {
+	if s.closed() {
+		return nil, ErrEngineClosed
+	}
+
 	if influxql.Sources(opt.Sources).HasSystemSource() {
 		// Only support a single system source.
 		if len(opt.Sources) > 1 {
@@ -483,8 +509,6 @@ func (s *Shard) SeriesKeys(opt influxql.IteratorOptions) (influxql.SeriesList, e
 		return []influxql.Series{{Aux: auxFields}}, nil
 	}
 
-	s.mu.RLock()
-	defer s.mu.RUnlock()
 	return s.engine.SeriesKeys(opt)
 }
 

--- a/tsdb/shard_test.go
+++ b/tsdb/shard_test.go
@@ -22,8 +22,6 @@ import (
 const DefaultPrecision = "s"
 
 func TestShardWriteAndIndex(t *testing.T) {
-	t.Skip("pending tsm1 iterator impl")
-
 	tmpDir, _ := ioutil.TempDir("", "shard_test")
 	defer os.RemoveAll(tmpDir)
 	tmpShard := path.Join(tmpDir, "shard")
@@ -34,6 +32,13 @@ func TestShardWriteAndIndex(t *testing.T) {
 	opts.Config.WALDir = filepath.Join(tmpDir, "wal")
 
 	sh := tsdb.NewShard(1, index, tmpShard, tmpWal, opts)
+
+	// Calling WritePoints when the engine is not open will return
+	// ErrEngineClosed.
+	if got, exp := sh.WritePoints(nil), tsdb.ErrEngineClosed; got != exp {
+		t.Fatalf("got %v, expected %v", got, exp)
+	}
+
 	if err := sh.Open(); err != nil {
 		t.Fatalf("error opening shard: %s", err.Error())
 	}
@@ -92,8 +97,6 @@ func TestShardWriteAndIndex(t *testing.T) {
 }
 
 func TestShardWriteAddNewField(t *testing.T) {
-	t.Skip("pending tsm1 iterator impl")
-
 	tmpDir, _ := ioutil.TempDir("", "shard_test")
 	defer os.RemoveAll(tmpDir)
 	tmpShard := path.Join(tmpDir, "shard")
@@ -151,7 +154,18 @@ func TestShardWriteAddNewField(t *testing.T) {
 
 // Ensure a shard can create iterators for its underlying data.
 func TestShard_CreateIterator(t *testing.T) {
-	sh := MustOpenShard()
+	sh := NewShard()
+
+	// Calling CreateIterator when the engine is not open will return
+	// ErrEngineClosed.
+	_, got := sh.CreateIterator(influxql.IteratorOptions{})
+	if exp := tsdb.ErrEngineClosed; got != exp {
+		t.Fatalf("got %v, expected %v", got, exp)
+	}
+
+	if err := sh.Open(); err != nil {
+		t.Fatal(err)
+	}
 	defer sh.Close()
 
 	sh.MustWritePointsString(`

